### PR TITLE
[Chat] File Sharing Refactor

### PR DESCRIPTION
**Pull request by jpeng-ms**
# What

This PR contains the following changes:
- Split data model into download and upload specific by introducing `AttachmentMetadataWithProgress`
- Updated uploadOptions to include `handleAttachmentSelection` and `handleAttachmentRemoval`
- Deprecate `AttachmentUploadAdapter` by making it internal
- Renamed UploadManager to UploadTask

# Why

Address review feedback

# How Tested

CI

# Process & policy checklist

- [X] I have updated the project documentation to reflect my changes if necessary.
- [X] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [X] This change causes current functionality to break.

### CODE REVIEW

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationAttachmentUploadAdapter.ts [hunk 1]
```diff
@@ -65,7 +69,9 @@ class AttachmentUploadContext {
 
   public updateAttachmentUpload(
     id: string,
-    data: Partial<Pick<AttachmentMetadata, 'progress' | 'id' | 'name' | 'extension' | 'uploadError' | 'url'>>
+    data: Partial<
+      Pick<AttachmentMetadataWithProgress, 'progress' | 'id' | 'name' | 'extension' | 'uploadError' | 'url'>
```
Comment (emlynmac) [inline packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationAttachmentUploadAdapter.ts:73]: Do the list of pick strings need to be revised?

File: packages/react-components/src/components/SendBox.tsx [hunk 2]
```diff
@@ -163,11 +163,11 @@ export interface SendBoxProps {
   onRenderAttachmentUploads?: () => JSX.Element;
   /* @conditional-compile-remove(attachment-upload) */
   /**
-   * Optional array of active attachment uploads where each object has attributes
+   * Optional array of AttachmentMetadataWithProgress where each object has attributes
```
[the remark this answers is not in the document]
Comment (jpeng-ms) [reply packages/react-components/src/components/SendBox.tsx:166]: the previous comment refers to ActiveAttachmentUpload which is the old name before refactor

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationAttachmentUploadAdapter.ts (hunk 1 above)
Comment (jpeng-ms) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationAttachmentUploadAdapter.ts:73]: nope, unless we get new feedback to rename any properties

File: packages/react-components/src/components/AttachmentCard.tsx [hunk 3]
```diff
@@ -37,7 +37,7 @@ export interface _AttachmentCardProps {
   /**
    * Attachment details including name, extension, url, etc.
    */
-  attachment: AttachmentMetadata;
+  attachment: AttachmentMetadata | AttachmentMetadataWithProgress;
```
[the remark this answers is not in the document]
Comment (jpeng-ms) [reply packages/react-components/src/components/AttachmentCard.tsx:40]: no, this is the base component that's used for both upload and download, so it has to be accepting data model for both

File: packages/react-components/src/components/AttachmentCard.tsx [hunk 4]
```diff
@@ -61,9 +61,11 @@ export interface _AttachmentCardProps {
 export const _AttachmentCard = (props: _AttachmentCardProps): JSX.Element => {
   const { attachment, menuActions, onActionHandlerFailed } = props;
   const attachmentCardStyles = useAttachmentCardStyles();
+
   const progress = useMemo(() => {
-    return attachment.progress;
-  }, [attachment.progress]);
+    return (attachment as AttachmentMetadataWithProgress).progress;
```
[the remark this answers is not in the document]
Comment (jpeng-ms) [reply packages/react-components/src/components/AttachmentCard.tsx:66]: the attachment card is shared for both upload and download, so at component level we don't know the use case until we check for the type of data model

Comment (emlynmac) [inline packages/react-components/src/components/AttachmentCard.tsx:66]: Does this work?
```suggestion
    return 'progress' in attachment ? attachment.progress : undefined;
```

File: packages/react-components/src/types/Attachment.ts [hunk 5]
```diff
@@ -27,6 +27,14 @@ export interface AttachmentMetadata {
    * Download URL for the attachment.
    */
   url?: string;
+}
+
+/**
+ * Metadata containing basic information about the uploaded attachment.
```
[the remark this answers is not in the document]
Comment (jpeng-ms) [reply packages/react-components/src/types/Attachment.ts]: thanks updated

File: packages/react-components/src/components/AttachmentCard.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (jpeng-ms) [reply packages/react-components/src/components/AttachmentCard.tsx]: thanks updated

File: packages/react-components/src/components/SendBox.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpeng-ms) [reply packages/react-components/src/components/SendBox.tsx]: thanks added `@link`

File: packages/communication-react/review/beta/communication-react.api.md [hunk 6]
```diff
@@ -1677,7 +1685,7 @@ export type CaptionsReceivedListener = (event: {
 }) => void;
 
 // @public
-export type ChatAdapter = ChatAdapterThreadManagement & AdapterState<ChatAdapterState> & Disposable_2 & ChatAdapterSubscribers & AttachmentUploadAdapter;
+export type ChatAdapter = ChatAdapterThreadManagement & AdapterState<ChatAdapterState> & Disposable_2 & ChatAdapterSubscribers & _AttachmentUploadAdapter;
```
Comment (emlynmac) [inline packages/communication-react/review/beta/communication-react.api.md:1688]: Should the adapter still be here?

File: packages/communication-react/review/beta/communication-react.api.md [hunk 7]
```diff
@@ -1724,7 +1733,7 @@ export interface ChatAdapterThreadManagement {
 // @public
 export type ChatAdapterUiState = {
     error?: Error;
-    attachmentUploads?: AttachmentUploadsUiState;
+    attachmentUploads?: _AttachmentUploadsUiState;
```
Comment (emlynmac) [inline packages/communication-react/review/beta/communication-react.api.md:1736]: Should this be gone too?

File: packages/react-components/src/components/AttachmentUploadCards.tsx [hunk 8]
```diff
@@ -29,10 +29,10 @@ export interface _AttachmentUploadCardsStrings {
  */
 export interface AttachmentUploadCardsProps {
   /**
-   * Optional array of active attachment uploads where each object has attibutes
+   * Optional array of AttachmentMetadataWithProgress where each object has attibutes
```
Comment (emlynmac) [inline packages/react-components/src/components/AttachmentUploadCards.tsx:32]:
```suggestion
   * Optional array of {@link AttachmentMetadataWithProgress} where each object has attributes
```

File: packages/react-composites/src/composites/CallWithChatComposite/state/CallWithChatAdapterState.ts [hunk 9]
```diff
@@ -40,11 +40,11 @@ export interface CallWithChatAdapterUiState {
   /**
    * Files being uploaded by a user in the current thread.
    * Should be set to null once the upload is complete.
-   * Array of type {@link AttachmentUploadsUiState}
+   * Array of type {@link _AttachmentUploadsUiState}
    *
    * @beta
    */
-  attachmentUploads?: AttachmentUploadsUiState;
+  attachmentUploads?: _AttachmentUploadsUiState;
```
Comment (emlynmac) [inline packages/react-composites/src/composites/CallWithChatComposite/state/CallWithChatAdapterState.ts:47]: What needs to change here- visibility or naming?

Comment (jpeng-ms) [reply packages/react-composites/src/composites/CallWithChatComposite/state/CallWithChatAdapterState.ts:47]: visibility, updated thanks

File: packages/communication-react/review/beta/communication-react.api.md (hunk 7 above)
Comment (jpeng-ms) [reply packages/communication-react/review/beta/communication-react.api.md:1736]: yes it should

File: packages/communication-react/review/beta/communication-react.api.md (hunk 6 above)
Comment (jpeng-ms) [reply packages/communication-react/review/beta/communication-react.api.md:1688]: yes it needs to be there until we refactor internal logic to remove upload adapter completely after the beta release

File: packages/communication-react/review/beta/communication-react.api.md [hunk 10]
```diff
@@ -1262,7 +1270,7 @@ export interface CallWithChatAdapterSubscriptions {
 // @public
 export interface CallWithChatAdapterUiState {
     // @beta
-    attachmentUploads?: AttachmentUploadsUiState;
+    attachmentUploads?: _AttachmentUploadsUiState;
```
[the remark this answers is not in the document]
Comment (jpeng-ms) [reply packages/communication-react/review/beta/communication-react.api.md]: this property should be internal, updated, thanks

**Review by jimchou-dev — APPROVED** [review]

File: packages/react-components/src/components/SendBox.tsx [hunk 11]
```diff
@@ -163,11 +163,11 @@ export interface SendBoxProps {
   onRenderAttachmentUploads?: () => JSX.Element;
   /* @conditional-compile-remove(attachment-upload) */
   /**
-   * Optional array of active attachment uploads where each object has attributes
-   * of a attachment upload like name, progress, errorMessage etc.
+   * Optional array of type {@linkAttachmentMetadataWithProgress}
```
Comment (emlynmac) [inline packages/react-components/src/components/SendBox.tsx:166]:
```suggestion
   * Optional array of type {@link AttachmentMetadataWithProgress}
```

Comment (emlynmac) [inline packages/react-components/src/components/SendBox.tsx:166]:
```suggestion
   * Optional array of type {@link AttachmentMetadataWithProgress}
```

File: packages/react-components/src/types/Attachment.ts [hunk 12]
```diff
@@ -4,7 +4,7 @@
 import { ChatMessage } from './ChatMessage';
 
 /**
- * Metadata containing basic information about the uploaded attachment.
+ * Metadata containing basic information about attachment.
```
Comment (emlynmac) [inline packages/react-components/src/types/Attachment.ts:7]:
```suggestion
 * Metadata containing basic information about the attachment.
```

Comment (emlynmac) [inline packages/react-components/src/types/Attachment.ts:7]:
```suggestion
 * Metadata containing basic information about the attachment.
```

Comment (emlynmac) [inline packages/react-components/src/types/Attachment.ts:7]:
```suggestion
 * Metadata containing basic information about the attachment.
```

**Review by emlynmac — APPROVED** [review]

**Pull request merged**